### DISCUSSION TITLE
Add support for Fronius Inverter

### DISCUSSION THREAD

**Discussion by slipx06** [slipx06/sunsynk-power-flow-card#368] [Fronius]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Add support for Fronius Inverters

### Expected behaviour

Add inverter status codes and image

### Mode

Both

### Context / Reason

Add additional inverters

**Comment by slipx06** [comment]
List of status codes 
[Fronius Status Codes.xlsx](https://github.com/slipx06/sunsynk-power-flow-card/files/13559464/Fronius.Status.Codes.xlsx)

Image
![Fronius](https://github.com/slipx06/sunsynk-power-flow-card/assets/7227275/66838408-8f03-46f5-b475-143fe7b91cc3)

**Comment by wgumaa** [comment]
Working on it :)

**Comment by wgumaa** [comment]
In addition, some of the sensors I used to get it to work have been made into templates. Here are some of the conversions: 

`    convert_grid_power:
      unique_id: convert_grid_power
      friendly_name: "Grid Power Watts"
      value_template: "{{ states('sensor.solarnet_power_grid') | float * 1000 | round(2) }}"
      
    convert_total_solar:
      unique_id: convert_total_solar
      friendly_name: "Total Solar Year KW"
      value_template: "{{ states('sensor.symo_advanced_15_0_3_m_1_energy_year') | float / 1000 | round(2) }}"
      
    convert_power_photovoltaics:
      unique_id: convert_power_photovoltaics
      friendly_name: "Power Photovoltaics W"
      value_template: "{{ states('sensor.solarnet_power_photovoltaics') | float(0) * 1000 | round(2) }}"
      
    convert_solar_day:
      unique_id: convert_solar_day
      friendly_name: "Solar Day W"
      value_template: "{{ states('sensor.symo_advanced_15_0_3_m_1_energy_days') | float * 1000 | round(2) }}"
      
    convert_charge_power:
      unique_id: convert_charge_power
      friendly_name: "Charge Power KW"
      value_template: "{{ (states('sensor.wattpilot_charging_energy') | float / 1000) | round(2) }}" 
      unit_of_measurement: kWh

    house_daily_load:
      unique_id: house_daily_load
      friendly_name: "House Daily Load"
      value_template: "{{ (states('sensor.daily_energy_import') | float + states('sensor.solarnet_energy_day') | float - states('sensor.convert_daily_energy_export') | float )}}" 
      unit_of_measurement: kWh

    essential_load:
      unique_id: essential_load
      friendly_name: "Essential Load"
      value_template: "{{ (states('sensor.symo_advanced_15_0_3_m_1_ac_power') | float(0) + states('sensor.convert_grid_power') | float )}}" 
      unit_of_measurement: kWh
      
    convert_daily_energy_export:
      unique_id: convert_daily_energy_export
      friendly_name: "Daily Energy Export KW"
      value_template: "{{ (states('sensor.daily_energy_export') | float / 1000) | round(2) }}"
      unit_of_measurement: kWh
      
    house_average_voltage:
      unique_id: house_average_voltage
      friendly_name: "House Average Voltage"
      value_template: "{{ ((states('sensor.smart_meter_ts_65a_3_voltage_phase_1') | float + states('sensor.smart_meter_ts_65a_3_voltage_phase_2') | float + states('sensor.smart_meter_ts_65a_3_voltage_phase_3') | float ) /3) | round(2) }}"
      `

**Comment by slipx06** [comment]
You don't need to convert W to kW or kWh to Wh and visa versa anymore. If you have defined the correct unit of measurement and have `auto_scale` enabled it will calculate the conversion for you.

**Comment by slipx06** [comment]
@wgumaa I've added support for the image in 4.7.0

**Comment by wgumaa** [comment]
Thanks @slipx06. I am away for the holidays but when I get back I will finish the mapping and update you with the entities to include the fronius system.

**Comment by wgumaa** [comment]
[Fronius Status Codes.xlsx](https://github.com/slipx06/sunsynk-power-flow-card/files/14458352/Fronius.Status.Codes.xlsx)

Here is the mapping I did the best I could. As you said too many codes!

**Comment by wgumaa** [comment]
So the Fronius is made up of 3 components: 

1. Smart Meter TS 65A-3
2. SolarNet Data Manager 
3. Symo Advanced 15.0-3-M - Inverter

**Fronius Smart Meter:**

![image](https://github.com/slipx06/sunsynk-power-flow-card/assets/105917603/5dea0589-1c09-496d-94f9-a24ce7f505f2)

**SolarNet Data Manager:**

![image](https://github.com/slipx06/sunsynk-power-flow-card/assets/105917603/7bc4f22e-9e25-4df3-9710-4927c62bde6d)

**Symo Advanced 15.0-3-M - Inverter**

![image](https://github.com/slipx06/sunsynk-power-flow-card/assets/105917603/186de89f-c2b4-40e2-b6d2-9a2091c78c68)


This my current view: 

![image](https://github.com/slipx06/sunsynk-power-flow-card/assets/105917603/fc31f4e2-56c1-4fed-82e7-8dff300bc679)

You can see that there are some figures that are completely off or not working: 

1. Daily load
2. Not sure that the autarky value is correct
3. and the load going into the house is showing minus

**Comment by wgumaa** [comment]
Here is the latest: 

![image](https://github.com/slipx06/sunsynk-power-flow-card/assets/105917603/dfdc7d85-7e86-4890-bf17-187255c11294)

I fixed the daily load but reinstating the template sensor: 

`house_daily_load:
  unique_id: house_daily_load
  friendly_name: "House Daily Load"
  value_template: "{{ (states('sensor.daily_energy_import') | float + states('sensor.solarnet_energy_day') | float - states('sensor.convert_daily_energy_export') | float )}}" 
  unit_of_measurement: kWh`

But the load going in is still wrong.

**Comment by slipx06** [comment]
The load is being calculated base on your L1 +L2 +L3 sensors. 

![image](https://github.com/slipx06/sunsynk-power-flow-card/assets/7227275/459d0125-3bb9-4c7f-af17-d054be6598b6)

It look like you have a sensor from the SolarNet Data Manager that you can use for  Essential, namely  `Power Load`

So try adding 

`essential_power: sensor.power_load` <----not sure the exact name

**Comment by wgumaa** [comment]
Yes there is a sensor "sensor.solarnet_power_load" Yes but still showing a negative value. Shouldn't it be positive? 

![image](https://github.com/slipx06/sunsynk-power-flow-card/assets/105917603/44789067-c5ef-4fcd-81d6-ec2ac0e29ab1)

Considering that the heatpump and charger are pulling? 

![image](https://github.com/slipx06/sunsynk-power-flow-card/assets/105917603/a029ed3e-bfde-4d4e-a4f3-1cf2e86218d8)

**Comment by wgumaa** [comment]
The last update broke the ratio and autarky

4.18.1 

![image](https://github.com/slipx06/sunsynk-power-flow-card/assets/105917603/6422095d-97ec-4843-a0d8-236ed772c892)

**Comment by slipx06** [comment]
Its working as intended. You cant have negative Autarky. Your essential load is -2.36kW which is the problem

**Comment by wgumaa** [comment]
I see that but that is calculated and I can't change that unless I make a template sensor. How have you calculated the essential load?

**Comment by wgumaa** [comment]
ok we have to figure a way to calculate the essential power. Using "sensor.power_load" is giving a negative value all the time. Is there a way to inverse that through a template?

**Comment by slipx06** [comment]
Yes you can create a template sensor that simply multiplies by -1. I don't have an option in the card to invert the load at the moment, but maybe its something I can add in the future

```
inverted_power_load:
  unique_id: inverter_power_load
  friendly_name: "Inverter Power Load"
  value_template: "{{ (states('sensor.power_load') | float * -1)}}" 
  unit_of_measurement: W
```

**Comment by slipx06** [comment]
I've added an option to invert the load power in https://github.com/slipx06/sunsynk-power-flow-card/releases/tag/v4.19.0

**Comment by wgumaa** [comment]
The Inverse essential load seems to have done the trick! 

![image](https://github.com/slipx06/sunsynk-power-flow-card/assets/105917603/c3979059-e509-4168-b643-c1ccdb516091)

**Comment by wgumaa** [comment]
Another bug appeared: when the system is running wholly on solar - not drawing anyhting from the grid, the Autarky and Ratio show 0. 

![image](https://github.com/slipx06/sunsynk-power-flow-card/assets/105917603/0e305e33-e0dc-434e-ba78-94f2591c6094)

What's interesting is when I plugged the car to charge, and the export energy dropped, the Autarky and Ratio now appear:

![image](https://github.com/slipx06/sunsynk-power-flow-card/assets/105917603/905a32af-750a-41e3-a450-5d1908bad5e5)

**Comment by wgumaa** [comment]
Could be because the energy going out is negative? 

I noticed it disappears as soon as it goes in KW but still there when in W

**Comment by slipx06** [comment]
Autarky only considers in your case the PV and Essential load. So there is something odd with the sensors. It wont work if your power to load is negative.  The Grid power has no influence on the calculation.

**Comment by slipx06** [comment]
@wgumaa I've gone though the list of status codes and these are my suggested mappings. These moslty map to error or fault conditions. I have confirmed here https://goodhewsolar.com.au/fronius-fault-codes/
I'm not seeing many "normal" or "standby" states. Can you check

<html


<body link="#0563C1" vlink="#954F72">

Code | Result | Status
-- | -- | --
  |   |  
101 | Grid voltage beyond permitted limits | Alarm
102 | AC voltage is too high | Alarm
103 | AC voltage is too low | Alarm
104 | Mains frequency beyond permitted limits | Alarm
105 | AC frequency is too high | Alarm
106 | AC frequency is too low | Alarm
107 | AC is outside the allowed thresholds | Alarm
108 | Independent operation detected | Alarm
112 | A residential current monitoring unit (RCMU) error has occurred | Fault
141 | Grid frequency increased above threshold | Alarm
  |   |  
301 | Overcurrent (AC) | Alarm
302 | Overcurrent (DC) | Alarm
303 | DC module overheating | Alarm
304 | AC module overheating | Alarm
305 | No power being fed in, despite closed relay | Normal
306 | PV output too low for feeding energy into the grid | Normal
307 | Low PV Voltage | Normal
308 | Intermediate circuit voltage too high | Alarm
309 | DC input voltage MPPT 1 too high | Alarm
311 | Polarity of DC strings reversed | Alarm
313 | DC input voltage MPPT 2 too high | Alarm
314 | Current sensor calibration timeout | Alarm
315 | AC current sensor error | Fault
316 | Interrupt Check fail | Fault
325 | Overtemperature in the connection area | Alarm
326 | Fan 1 error | Fault
327 | Fan 2 error | Fault
  |   |  
401 | No communication possible with the power stage set | Alarm
406 | AC module temperature sensor faulty (L1) | Fault
407 | AC module temperature sensor faulty (L2) | Fault
408 | DC component measured in the grid too high | Alarm
412 | Fixed voltage mode has been selected instead of MPP voltage mode |  Fault
415 | Safety cut out via option card or RECERBO has triggered |  Fault
416 | No communication possible between power stage set and control system | Fault
417 | Hardware ID problem | Fault
419 | Unique ID conflict | Fault
420 | No communication possible with the Hybrid manager | Fault
421 | HID range error | Fault
425 | No communication with the power stage set possible | Fault
426 | Possible hardware fault | Fault
427 | Possible hardware fault | Fault
428 | Possible hardware fault | Fault
431 | Software problem | Fault
436 | Functional incompatibility | Fault
437 | Power stage set problem | Fault
438 | Functional incompatibility | Fault
443 | Intermediate circuit voltage too low or asymmetric | Fault
445 | Compatability error | Fault
447 | Insulation fault | Fault
448 | Neutral conductor not connected | Alarm
450 | Guard cannot be found | Alarm
451 | Memory error detected | Fault
452 | Communication error between the processors | Fault
453 | Grid voltage and power stage set are incompatible | Fault
454 | Grid frequency and power stage set are incompatible | Fault
456 | Anti-islanding function is no longer implemented correctly | Alarm
457 | Grid relay sticking or the neutral conductor ground voltage is too high | Fault
458 | Error when recording the measuring signal | Fault
459 | Error when recording the measuring signal for insulation test | Fault
460 | Reference voltage source for the DSP is working out of tolerance | Fault
461 | Error in the DSP data memory | Fault
462 | Fault with the DC feed monitoring routine | Fault
463 | Reversed AC polarity | Alarm
474 | The RCMU sensor is faulty | Fault
475 | Insulation fault | Fault
476 | The driver supply voltage is too low | Alarm
479 | Intermediate circuit relay has switched off | Alarm
480 | Functional incompatibility | Alarm
481 | Functional incompatibility | Alarm
482 | The setup phase after the initial start-up was interrupted. | Alarm
483 | The voltage UDC fixed on MPP2 string is out of limits | Alarm
485 | The CAN transmit buffer is full | Alarm
489 | Permanent intermediate circuit capacitor overvoltage | Alarm
  |   |  
502 | Insulation fault on the solar panel | Fault
509 | No energy has been fed into the grid in the last 24 hours | Normal
515 | Unable to communicate with filter | Alarm
516 | Unable to communicate with storage unit | Fault
517 | Overheating which is causing power derating | Alarm
518 | Internal DSP fault | Fault
519 | Unable to communicate with storage unit | Alarm
520 | No energy has been fed into the grid by MPPT1 in the past 24 hours | Normal
522 | The DC is low on string 1 | Normal
523 | The DC is low in string 2 | Normal
528 | There is a functional incompatibility | Alarm
529 | There is a functional incompatibility | Alarm
560 | The grid frequency has become excessively high | Alarm
564 | There is a functional incompatibility | Alarm
566 | The arc detector has been deactivated | Alarm
567 | The GVDPR is active and the inverter will operate with reduced power   output | Alarm
  |   |  
601 | The CAN bus is full | Alarm
603 | The AC module sensor is faulty | Fault
604 | The DC module sensor is faulty | Fault
607 | RCMU fault | Fault
608 | There is a functional incompatibility | Fault
</body>
</html>

**Reply by wgumaa** [reply]
Yes the list looks complete.

I went into the entity: 

![image](https://github.com/slipx06/sunsynk-power-flow-card/assets/105917603/7a4db005-2cbe-4a4a-a11f-d1f8112f0603)

**Reply by wgumaa** [reply]
Where its showing status codes, there are various numbers: 7, 2, 3 but there is no details as to what they are.

**Reply by wgumaa** [reply]
As for error codes: 

0 - all good
I only get 2 error codes currently:

306: PV output too low for feeding energy into the grid
307: Low PV Voltage

Looking at the history, these codes occur at night when the sun has set.

**Reply by wgumaa** [reply]
Look here also for codes

https://pswenergy.com.au/fronius-inverter-state-codes/

**Reply by slipx06** [reply]
@wgumaa I see that you have both `error code` and `status code` sensors? I assume that the table above is only for the error codes and we don't really want to use that for the inverter status sensor in the card. Please can you share a link to the home assistant integration you are using so we can try figure out what the status codes mean i.e 7,2,3 etc

**Reply by wgumaa** [reply]
Sure. It's part of the core:

https://www.home-assistant.io/integrations/fronius

**Reply by slipx06** [reply]
Found them Will use this for the mapping

```
_INVERTER_STATUS_CODES: Final[dict[int, InverterStatusCodeOption]] = {
    0: InverterStatusCodeOption.STARTUP,
    1: InverterStatusCodeOption.STARTUP,
    2: InverterStatusCodeOption.STARTUP,
    3: InverterStatusCodeOption.STARTUP,
    4: InverterStatusCodeOption.STARTUP,
    5: InverterStatusCodeOption.STARTUP,
    6: InverterStatusCodeOption.STARTUP,
    7: InverterStatusCodeOption.RUNNING,
    8: InverterStatusCodeOption.STANDBY,
    9: InverterStatusCodeOption.BOOTLOADING,
    10: InverterStatusCodeOption.ERROR,
    11: InverterStatusCodeOption.IDLE,
    12: InverterStatusCodeOption.READY,
    13: InverterStatusCodeOption.SLEEPING,
    255: InverterStatusCodeOption.UNKNOWN,
```

**Reply by slipx06** [reply]
```
        standby: {states: ['8', '11', '13'], color: 'blue', message: localize('common.standby')},
        selftest: {states: ['0', '1', '2', '3', '4', '5', '6', '9'], color: 'yellow', message: localize('common.selftest')},
        normal: {states: ['7', '12'], color: 'green', message: localize('common.normal')},
        alarm: {states: ['3'], color: 'orange', message: localize('common.alarm')},
        fault: {states: ['255'], color: 'red', message: localize('common.fault')},
```

**Comment by slipx06** [comment]
https://github.com/slipx06/sunsynk-power-flow-card/releases/tag/v4.29.0 let me know if it works

**Comment by cloudbr34k84** [comment]
would anyone mind sharing their Fronius card yaml code. Having a hard time figuring out where everything goes?

**Reply by wgumaa** [reply]
> would anyone mind sharing their Fronius card yaml code. Having a hard time figuring out where everything goes?
> 
> 

Sure. Give me some time. Not at home right now

**Reply by wgumaa** [reply]
@cloudbr34k84 which inverter are you using?

**Reply by wgumaa** [reply]
Can't DM you here. Are you on the HA forums? What's your @

**Reply by cloudbr34k84** [reply]
its @cloudbr34k on the HA community forums
Im using 3 phase Symo 10.0-3-M

**Reply by wgumaa** [reply]
Just sent!

**Reply by sausagedogsalad** [reply]
> Just sent!

@wgumaa @cloudbr34k84 I am having issues working out which entities go where in the card - I am running entities from MPPT, Fronius Integration and SolarNet. I have a Primo 10kw, and smart meter on a 3 phase. Have you published your YAML anywhere - mostly interested in what correlates to what in the card? Thanks

**Reply by wgumaa** [reply]
@sausagedogsalad Hi, 

I am currently using the compact view and my card looks like this: 

![image](https://github.com/user-attachments/assets/0183e3e2-71f1-451a-8ec4-87f83effe5c0)

The YAML code is: 

`type: custom:sunsynk-power-flow-card
cardstyle: compact
title: null
title_size: 16px
title_colour: green
panel_mode: true
card_height: null
show_solar: true
show_battery: false
decimal_places: 2
inverter:
  modern: false
  model: fronius
  colour: green
  autarky: energy
  auto_scale: true
  three_phase: true
solar:
  colour:
    - 225
    - 169
    - 16
  display_mode: 2
  show_daily: true
  mppts: 1
  pv1_name: Solar
  max_power: 12000
  auto_scale: true
  dynamic_colour: true
load:
  show_aux: false
  show_daily_aux: false
  additional_loads: 3
  show_daily: true
  auto_scale: true
  max_power: 13000
  load1_icon: mdi:heat-pump-outline
  load2_icon: mdi:ev-plug-type2
  three_phase: true
  dynamic_colour: true
  dynamic_icon: true
  invert_load: true
  load3_icon: mdi:fan
  load1_name: Heat
  load2_name: EV
  load3_name: Vent
grid:
  show_daily_buy: true
  show_daily_sell: true
  show_nonessential: false
  auto_scale: true
  export_colour:
    - 225
    - 169
    - 16
  colour:
    - 24
    - 158
    - 216
  invert_grid: false
  show_absolute: false
  invert_flow: false
entities:
  grid_connected_status_194: null
  grid_power_169: sensor.convert_grid_power
  day_pv_energy_108: sensor.solarnet_energy_day
  remaining_solar: sensor.energy_production_today_remaining
  essential_load1: sensor.instantaneous_used_power_32167
  essential_load2: sensor.wattpilot_charging_energy
  load_power_L1: sensor.smart_meter_ts_65a_3_real_power_phase_1
  load_power_L2: sensor.smart_meter_ts_65a_3_real_power_phase_2
  load_power_L3: sensor.smart_meter_ts_65a_3_real_power_phase_3
  environment_temp: sensor.current_outdoor_temperature_bt1_30002
  pv1_power_186: sensor.solarnet_power_photovoltaics
  total_pv_generation: sensor.symo_advanced_15_0_3_m_1_energy_year
  grid_ct_power_172: sensor.solarnet_power_grid
  day_load_energy_84: sensor.house_daily_load
  pv1_current_110: sensor.symo_advanced_15_0_3_m_dc_current
  pv1_voltage_109: sensor.symo_advanced_15_0_3_m_dc_voltage
  inverter_current_164: sensor.symo_advanced_15_0_3_m_ac_current
  inverter_voltage_154: sensor.symo_advanced_15_0_3_m_ac_voltage
  load_frequency_192: sensor.symo_advanced_15_0_3_m_frequency
  inverter_power_175: sensor.symo_advanced_15_0_3_m_ac_power
  essential_power: sensor.solarnet_power_load
  essential_load3: sensor.ventilation_current_power_usage
  day_grid_export_77: sensor.daily_energy_produced
  day_grid_import_76: sensor.daily_energy_consumed
large_font: true
show_grid: true
layout_options:
  grid_columns: 4
  grid_rows: auto
card_mod:
  style: |
    ha-card { 
      --ha-card-background: transparent;
      --ha-card-border-radius: 0px;
      --ha-card-box-shadow: 0px ;
      --ha-card-border-width: 0px;
    }
`

You can ignore the layout and card mod parts at the end and adjust your essential loads accordingly. 

You have to create 2 utility meters which i have named: 

Daily Energy Consumed - sensor.daily_energy_consumed
Daily Energy Produced - sensor.daily_energy_produced

I hope this helps.

**Reply by wgumaa** [reply]
@sausagedogsalad if you are using the full card do share your progress as i am too having some challenges with configuring that.

**Reply by sausagedogsalad** [reply]
@wgumaa thanks very much - this gives me a lot to go off. I am using the full card so will see how I go and report back. Thanks again.

**Reply by wgumaa** [reply]
Cheers do keep me posted!
